Add Llama Aave Treasury Tables - Eth

MSilb7 opened the pull request:
Adding 3 tables and inserts to store tables for Llama's Aave Dashboard: https://dune.xyz/llama/Aave-Treasury-Finances-by-Llama (timeouts are common, storing these abstraction tables is an attempt to speed this up versus building views on top of views).

All queries have been tested. See the readme for test queries (hopefully they all work).

I've checked that:

* [x ] the query produces the intended results
* [ ] the folder name matches the schema name
* [ ] the schema name exists in Dune
* [ ] views are prefixed with `view_`, functions with `fn_`.
* [ ] the filename matches the defined view, table or function and ends with .sql
* [ ] each file has only one view, table or function defined  
* [ ] column names are `lowercase_snake_cased`

aalan3 commented on ethereum/llama/aave_treasury/aave_daily_atoken_balances.sql:
```diff
@@ -0,0 +1,15 @@
+	starting_balance numeric,
+	interest_rate_apr numeric,
+	int_earned numeric,
+	total_bal numeric,
```
Any reason for shortening `total_bal`, could be `total_balance` otherwise?

MSilb7 replied:
updated

aalan3 commented on ethereum/llama/aave_treasury/aave_daily_atoken_balances.sql:
```diff
@@ -0,0 +1,15 @@
+	daily_change numeric,
+	starting_balance numeric,
+	interest_rate_apr numeric,
+	int_earned numeric,
```
`int_earned` -> `interest_earned`?

MSilb7 replied:
updated

aalan3 commented on ethereum/llama/aave_treasury/aave_fees_by_day.sql:
```diff
@@ -0,0 +1,21 @@
+CREATE SCHEMA IF NOT EXISTS llama;
+
+CREATE TABLE IF NOT EXISTS llama.aave_fees_by_day (   
```
`aave_daily_fees`? To keep it in line with the above names.

aalan3 commented on ethereum/llama/aave_treasury/aave_tokens.sql:
```diff
@@ -0,0 +1,15 @@
+    token_address bytea PRIMARY KEY,
+    decimals int4,
+    symbol text,
+    erc20_address bytea,
```
Is this the same as `underlying_token`?

MSilb7 replied:
conceptually, yes. Maybe that's a better name.

aalan3 commented on ethereum/llama/aave_treasury/aave_tokens.sql:
```diff
@@ -0,0 +1,15 @@
+CREATE SCHEMA IF NOT EXISTS llama;
+
+CREATE TABLE IF NOT EXISTS llama.aave_tokens (   
```
Is this table the same as this view? https://github.com/duneanalytics/abstractions/blob/master/ethereum/aave/view_atokens.sql

MSilb7 replied:
similar, but that one is static, and misses some context on each token. This should update for every live token with additional context (i.e. borrow v deposit, ARC v normal)

aalan3 commented on ethereum/llama/aave_treasury/insert_aave_daily_treasury_events.sql:
```diff
@@ -0,0 +1,275 @@
+    )
+	
+WITH addresses AS (
+SELECT address FROM dune_user_generated.llama_treasury_addresses WHERE protocol = 'Aave' AND blockchain = 'Ethereum'
```
things that live in `dune_user_generated` could be overwritten or removed. If this depends on in, can we make this an abstraction instead?

aalan3 commented on ethereum/llama/aave_treasury/insert_aave_daily_treasury_events.sql:
```diff
@@ -0,0 +1,275 @@
+                        WHERE "blockchain" = 'Ethereum' AND "protocol" = 'Aave')
+    AND "success" = false
+    AND block_time >= start_time_day AND block_time <= end_time_day
+    --GROUP BY 1,2,3
```
drop this line?

aalan3 commented on ethereum/llama/aave_treasury/insert_aave_daily_treasury_events.sql:
```diff
@@ -0,0 +1,275 @@
+    AND "tx_success" = true
+    AND block_time >= start_time_day AND block_time <= end_time_day
+    
+    --GROUP BY 1,2,3
```
drop this line?

aalan3 commented on ethereum/llama/aave_treasury/insert_aave_daily_treasury_events.sql:
```diff
@@ -0,0 +1,275 @@
+FROM (
+SELECT
+COALESCE(m.address,a."contract_address") AS contract_address,  --handle for token migrations
+--evt_tx_hash,--counterparty,
```
Drop this line and the line below?

aalan3 commented on ethereum/llama/aave_treasury/insert_aave_daily_treasury_events.sql:
```diff
@@ -0,0 +1,275 @@
+    CASE WHEN t."to" IN (SELECT address FROM addresses) THEN 'Transfer Out'
+    WHEN t."evt_tx_hash" IN (SELECT "evt_tx_hash" FROM aave."stkAAVE_evt_Staked"
+                                WHERE "from" IN (SELECT address FROM addresses) ) THEN 'Staking Out'
+    /*WHEN t."to" IN ('\xa1116930326D21fB917d5A27F1E9943A9595fb47', --Balancer Pool
```
Uncomment or drop?

aalan3 commented on ethereum/llama/aave_treasury/insert_aave_tokens.sql:
```diff
@@ -0,0 +1,106 @@
+    FROM llama.aave_tokens
+);
+
+-- tables are so small, so just run everything rather than building in logic for backfilling tokens. Set a 1 year gap to add some constraint.
```
If this table is similar to https://github.com/duneanalytics/abstractions/blob/master/ethereum/aave/view_atokens.sql, maybe this could be a materialized view instead?

aalan3 commented on ethereum/llama/aave_treasury/readme.md:
```diff
@@ -0,0 +1,17 @@
+# Llama Aave Treasury Dashboard Tables
```
Thanks for adding a readme ⭐️⭐️

aalan3 commented on ethereum/llama/aave_treasury/readme.md:
```diff
@@ -0,0 +1,17 @@
+
+----
+
+### Testing Information
```
⭐️

aalan3 reviewed (commented):
Awesome work! I had some comments and suggestions, let me know what you think.

MSilb7 reviewed (commented):
Making some naming updates and other changes to incorporate Aave's ARC and Real-World Asset programs, will comment again when it's good to go

MSilb7 commented:
I believe all of these updates are in now. Main highlights:
- Renamed some fields as per comments
- Moved tables under the Aave schema

MSilb7 commented:
Hi, following up to see if this is good to go?

aalan3 commented:
Apologies for the delay @MSilb7, I'll make sure to review tomorrow.

aalan3 commented on ethereum/aave/llama_aave_treasury/insert_aave_daily_treasury_events.sql:
```diff
@@ -0,0 +1,272 @@
+    )
+	
+WITH addresses AS (
+SELECT address FROM dune_user_generated.llama_treasury_addresses WHERE protocol = 'Aave' AND blockchain = 'Ethereum'
```
`dune_user_generated.llama_treasury_addresses` this should be its own table. Things `dune_user_generated` can be removed and then this would break.

(Same applies for all occurrences of this table)

aalan3 commented on ethereum/aave/insert_aave_tokens.sql:
```diff
@@ -0,0 +1,169 @@
+l.contract_address AS token_address,
+bytea2numeric(substring(data from 65 for 32)) AS decimals,
+regexp_replace(REPLACE(encode(substring(data from 288 for 128),'escape')::text,'\000',''), '\W', ' ', 'g') AS symbol,
+dune_user_generated.get_address_from_data(topic2) AS underlying_token_address,
```
This should be added as a abstraction function, so it's not deleted.

MSilb7 replied:
Added the function in public to this PR

aalan3 reviewed (commented):
Almost there now! Just had a few comments on relying on `dune_user_generated`.

MSilb7 commented:
Updates should be in now!

- Added the function to get addresses from data in the public folder.
- Added the llama treasury addresses table in the llama schema.

MSilb7 commented:
Hi, bumping this back up again!

aalan3 commented on ethereum/aave/insert_aave_daily_atoken_balances.sql:
```diff
@@ -0,0 +1,136 @@
+COALESCE("interest_rate_apr",0) AS interest_rate_apr, drank
+FROM (
+SELECT pre.*,
+DENSE_RANK() OVER (PARTITION BY pre."contract_address" ORDER BY DATE_TRUNC('day',pre."evt_block_time") ASC) AS drank
```
This causes 
```
CREATE FUNCTION
psql:ethereum/aave/insert_aave_daily_atoken_balances.sql:127: ERROR:  column pre.evt_block_time does not exist
LINE 27: ... pre."contract_address" ORDER BY DATE_TRUNC('day',pre."evt_b...
                                                              ^
...                          
```

aalan3 replied:
Not sure if this should be `day` or if `evt_block_time` is missing below?

MSilb7 replied:
This should be day, just fixed it!

aalan3 reviewed (changes requested):
I ran into some issues when deploy this, mostly typos which I fixed and pushed. But added a comment on one thing that I wasn't completely sure how to fix.

MSilb7 commented:
Thanks so much for the detail + fixes. Just pushed an update for the day issue.

aalan3 commented on ethereum/aave/insert_aave_daily_atoken_balances.sql:
```diff
@@ -0,0 +1,136 @@
+CREATE OR REPLACE FUNCTION aave.insert_aave_daily_atoken_balances(start_time timestamptz, end_time timestamptz) RETURNS integer
```
```
CREATE FUNCTION
 insert_aave_daily_atoken_balances
-----------------------------------
                                 0
(1 row)

INSERT 0 1
```

Something went wrong here or is this correct?

aalan3 replied:
Ah might need to run something below first I guess.

MSilb7 replied:
Yeah aave."aave_tokens" should be first, then likely [insert_aave_daily_interest_rates](https://github.com/duneanalytics/abstractions/pull/700/files#diff-b6a8eae9914b24dcb8bac4a08754a7a6d947dc34e8d18b82d2ee1be3dff59c8f), then this

aalan3 commented:
I've done the aave part, could you check if it's correct?

aalan3 commented on ethereum/aave/llama_aave_treasury/insert_aave_daily_treasury_fees.sql:
```diff
@@ -0,0 +1,149 @@
+contract_address,
+ version
+
+    ON CONFLICT (day, contract_address) DO UPDATE SET
```
this is giving:
```
CREATE FUNCTION
psql:ethereum/aave/llama_aave_treasury/insert_aave_daily_treasury_fees.sql:140: ERROR:  ON CONFLICT DO UPDATE command cannot affect row a second time
HINT:  Ensure that no rows proposed for insertion within the same command have duplicate constrained values.
```

aalan3 replied:
Still getting 
```
psql:ethereum/aave/llama_aave_treasury/insert_aave_daily_treasury_fees.sql:140: ERROR:  there is no unique or exclusion constraint matching the ON CONFLICT specification
```

aalan3 commented on ethereum/aave/llama_aave_treasury/insert_aave_daily_treasury_events.sql:
```diff
@@ -0,0 +1,287 @@
+    SELECT * FROM eth_transfers
+    
+) a
+LEFT JOIN llama.llama_token_migrations m
```
This table doesn't exist, is it missing or should it point to something else?

aalan3 commented:
@MSilb7 wanted to bump this, added a couple of comments.

MSilb7 commented:
> @MSilb7 wanted to bump this, added a couple of comments.

Thanks, sorry for the delay. Made the updates (and recognized the missing reference table).

aalan3 commented:
Sorry for the wait. The constraint issue is still there, I also added a couple of commits which fix some issues. There is also some polygon stuff in here that should be dropped (will be added in the polygon PR I guess)

MSilb7 commented:
Is the conflict issue still on [insert_aave_daily_treasury_fees.sql](https://github.com/duneanalytics/abstractions/pull/700/files/11fb314301c79225ce8ec6a8e2856fa5b842839a#diff-01e545cd93a8ea4a676b4067153e2613ca671ff2095d3f7be4d51473e5de15f5)? I'm seeing all unique rows when I test locally. The end of the query is grouped by the conflict constraints, so it doesn't make sense for there to be any issues.

Was the create table updated?

aalan3 commented:
>Was the create table updated?

My bad, no it wasn't. Looks like it's working now. I'm deploying the tables currently. Will merge when it's done.

aalan3 commented:
@MSilb7 merged, thank you! Could you take a look on the data as a final sanity check? And then also just have another look on the polygon version to make sure we don't make the same mistakes as here, and I'll deploy that one too.
